### ISSUE TITLE
400 Bad request / auth Invalid session state

### ISSUE THREAD

**Issue by DeLoWaN** [vouch/vouch-proxy#315]
Hello, I encouter a strange issue:

I configured multiple domains with Vouch with Google auth. It works... partially. I'm on the 0.17.3 version of Vouch.

I am testing in a browser with private mode.

First scenario, If I go directly to my protected page, it presents me a Google login page. I login with MFA, then I can see in dev bar that it cycles between my protected page / vouch-login / google-auth / vouch-auth.

After 4 cycles, it presents me an Error page with "too many redirects", then straight away, it redirects me to the protected page, fully logged and working.

Second scenario, I logged in first to my gmail account. I then try to connect my protected page. It does the same cycles than in the first scenario and ends up in a vouch page saying: `400 Bad request` and I have this error message in the logs : `/auth Invalid session state: stored %!s(<nil>)`

Third scenario, I log in into my Google account, and then I try to connect the protected page with the dev bar active and the `disable cache` checkbox ticked: works like a charm, no cycle.

I guess there is something wrong with the cookie ?

Here is my config: 

[config.yaml](https://pastebin.com/02E37Xah)

Important to know, I don't use my primary domain to connect the protected page, the protected page I'm testing is under `activity-logs.preprod.MYDOMAIN.fr`

[Nginx conf](https://pastebin.com/FNv5Pv8r) of my protected page.

[Nginx conf](https://pastebin.com/ym8D6CB2) of vouch proxy.

[log](https://pastebin.com/0cHVHSjB) of first scenario.

Can anyone help me with that ? Thanks a lot.

**Comment by bnfinet** [comment]
hello @kalumkalac 

Those `hasteb.in` links are empty :(.  Not sure if they timed out, could you please re-post with a `gist`?

What browser are you using?  Do you see the same behavior if you are not using the browser in Private Mode?

**Comment by DeLoWaN** [comment]
Hi @bnfinet , thank you for your interest. I edited my post, the links are working now.

**Comment by samarjoe** [comment]
I'm having the exact same intermittency of 400 Bad Request with
session state: stored %!s(<nil>)
also. 

I do notice that if I use some test page that is super lightweight, I don't get this though. Does anyone think it could be related to the size of the request?

**Comment by bnfinet** [comment]
@DeLoWaN Could you please post your logs in the manner described in the README.

What happens when you just use one domain, perhaps `preprod.MYDOMAIN.fr` in your VP config?

@samarjoe Could you please post your logs and config in the manner described in the README.  Have you tried increasing the size of your Nginx bufferes?

**Comment by DeLoWaN** [comment]
Ok, with further tests, I finally get it to work.

In my nginx config of the protected page (my php application), I had this instruction with was causing the issue:

```
more_set_headers 'Cache-Control: public, max-age=2592000, s-maxage=2592000';
```

Why ? I have no idea. But I soon as I disabled it, I had no longer these cycles and the scenario that was failing (2nd one) failed no more.

**Comment by bnfinet** [comment]
@DeLoWaN I'm glad that it's working now.

I don't see why the inclusion of that cache header would have bearing on the VP `/login` --> `/auth` cycle.  Did you change anything else?

Does your app include a path `/login` or `/auth`

Are you using CloudFlare or another proxy in front of your domain by any chance?  `s-maxage=2592000'` would affect the intermediary cache.

**Comment by DeLoWaN** [comment]
Yes, there was a bunch of others instructions, but, trying disabling them one by one, only this one lead to the bugs I encountered.

My app does not include these paths.

The app is a PHP-FPM and I linked earlier the conf at [this link](https://pastebin.com/FNv5Pv8r).

Was was missing from that conf though was an include of a conf file which contains (among others) headers instructions including the one I pointed out.

I'm am not using a CDN in front of my app, actually, this include conf came from another application which was using a CDN. I then realised I could get rid of that include since it was not needed in the protected page.
